File: backend/app/schemas/software.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


class SoftwareBase(BaseModel):
    name: str
    version: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    license_type: Optional[str] = None
    license_key: Optional[str] = None
    vendor: Optional[str] = None
    url: Optional[str] = None
    installed_on: Optional[str] = None
    status: str = "active"
    notes: Optional[str] = None


class SoftwareCreate(SoftwareBase):
    pass


class SoftwareUpdate(BaseModel):
    name: Optional[str] = None
    version: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    license_type: Optional[str] = None
    license_key: Optional[str] = None
    vendor: Optional[str] = None
    url: Optional[str] = None
    installed_on: Optional[str] = None
    status: Optional[str] = None
    notes: Optional[str] = None


class SoftwareResponse(SoftwareBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class SoftwareListResponse(BaseModel):
    items: list[SoftwareResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class SubscriptionBase(BaseModel):
    name: str
    provider: Optional[str] = None
    description: Optional[str] = None
    cost: Optional[float] = None
    billing_cycle: str = "monthly"
    start_date: Optional[datetime] = None
    expiry_date: Optional[datetime] = None
    auto_renew: int = 1
    category: Optional[str] = None
    url: Optional[str] = None
    status: str = "active"
    notes: Optional[str] = None


class SubscriptionCreate(SubscriptionBase):
    pass


class SubscriptionUpdate(BaseModel):
    name: Optional[str] = None
    provider: Optional[str] = None
    description: Optional[str] = None
    cost: Optional[float] = None
    billing_cycle: Optional[str] = None
    start_date: Optional[datetime] = None
    expiry_date: Optional[datetime] = None
    auto_renew: Optional[int] = None
    category: Optional[str] = None
    url: Optional[str] = None
    status: Optional[str] = None
    notes: Optional[str] = None


class SubscriptionResponse(SubscriptionBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class SubscriptionListResponse(BaseModel):
    items: list[SubscriptionResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/software.ts
export interface Software {
  id: number;
  name: string;
  version?: string;
  description?: string;
  category?: string;
  license_type?: string;
  license_key?: string;
  vendor?: string;
  url?: string;
  installed_on?: string;
  status: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface SoftwareCreate {
  name: string;
  version?: string;
  description?: string;
  category?: string;
  license_type?: string;
  license_key?: string;
  vendor?: string;
  url?: string;
  installed_on?: string;
  status?: string;
  notes?: string;
}

export type SoftwareUpdate = Partial<SoftwareCreate>;

export interface Subscription {
  id: number;
  name: string;
  provider?: string;
  description?: string;
  cost?: number;
  billing_cycle: string;
  start_date?: string;
  expiry_date?: string;
  auto_renew: number;
  category?: string;
  url?: string;
  status: string;
  notes?: string;
  created_at: string;
  updated_at: string;
}

export interface SubscriptionCreate {
  name: string;
  provider?: string;
  description?: string;
  cost?: number;
  billing_cycle?: string;
  start_date?: string;
  expiry_date?: string;
  auto_renew?: number;
  category?: string;
  url?: string;
  status?: string;
  notes?: string;
}

export type SubscriptionUpdate = Partial<SubscriptionCreate>;

export const LICENSE_TYPES = [
  "Perpetual",
  "Subscription",
  "Open Source",
  "Freeware",
  "Trial",
  "OEM",
  "Volume",
  "Other",
] as const;

export const BILLING_CYCLES = [
  { value: "monthly", label: "Monthly" },
  { value: "quarterly", label: "Quarterly" },
  { value: "yearly", label: "Yearly" },
  { value: "lifetime", label: "Lifetime" },
] as const;
